### PR TITLE
CLOUDP-411215: Move watchedResources setting to OperatorConfig

### PR DESCRIPTION
**Pull request by m1kola**
# Summary

Continues migrating operator settings into the `OperatorConfig` CR. Moves the set of CRDs the operator reconciles off the `-watch-resource` flag / `operator.watchedResources` Helm value onto `OperatorConfig.spec.watchedResources`, defaulting to all known CRDs. The previously implicit enablement of `mongodbmulticluster` (from `multiCluster.clusters`) and `clustermongodbroles` (from `operator.enableClusterMongoDBRoles`) is now expressed explicitly in that list. Member-cluster discovery now tolerates a missing member-list ConfigMap, so single-cluster operators start even with `mongodbmulticluster` watched.

Note: this branch also carries the commit from #1306 (voyageais catchup) — close #1306 in favour of this PR.

## Proof of Work

- `pkg/operatorconfig/operatorconfig_test.go` — defaulting (absent CR → all CRDs; explicit subset preserved)
- `helm template helm_chart` renders with no `-watch-resource` args across default / `multiCluster.clusters` / `enableClusterMongoDBRoles=false`
- e2e `operator_partial_crd` and `multicluster_search` migrated to `OperatorConfig`

## Checklist

- [x] Have you linked a jira ticket and/or is the ticket in the title?
- [x] Have you checked whether your jira ticket required DOCSP changes?
- [x] Have you added changelog file?
    - use `skip-changelog` label if not needed
    - refer to [Changelog files and Release Notes](https://github.com/mongodb/mongodb-kubernetes/blob/master/CONTRIBUTING.md#changelog-files-and-release-notes) section in CONTRIBUTING.md for more details

---

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -521,6 +498,12 @@ func getMemberClusters(ctx context.Context, cfg *rest.Config, currentNamespace s
 
 	m := corev1.ConfigMap{}
 	err = c.Get(ctx, types.NamespacedName{Name: util.MemberListConfigMapName, Namespace: currentNamespace}, &m)
+	if apierrors.IsNotFound(err) {
+		// No multi-cluster configuration present: run as single-cluster with no member clusters.
+		// The member-list ConfigMap is absent on single-cluster installs (e.g. when
+		// mongodbmulticluster is watched by default). Callers handle an empty member list.
+		return nil, nil
+	}
```
Comment (josvazg) [inline main.go:506]: > 1. main.go:501-506 — getMemberClusters swallowing NotFound conflates "not configured" with "misconfigured." This function only runs when mongodbmulticluster is watched — which is now the default case for every install, single- or multi-cluster. So a genuinely broken multi-cluster install (ConfigMap deleted, wrong namespace, RBAC returning a 404-shaped error) now silently degrades to "0 member clusters, single-cluster mode" with just a Warnf, instead of failing fast as it did before. Worth a second look: is there a way to distinguish "single-cluster, no multi-cluster CM expected" from "multi-cluster intended but the CM went missing"? Right now there isn't — both look identical.

**Comment by josvazg** [comment]
> 2. Real upgrade gotcha (documented, but sharp edge): existing OperatorConfig CRs with a restricted watchedResources subset don't get migrated. withDefaults only fills in AllWatchedResources when the field is empty, not merged into an existing non-empty list. So anyone who previously restricted resources via Helm (which never included mongodbmulticluster/clustermongodbroles unless separately opted in) and already has an OperatorConfig CR with an explicit subset will silently lose clustermongodbroles/mongodbmulticluster watching post-upgrade unless they manually update the CR first. The changelog does call this out explicitly ("must create or update their OperatorConfig resource before upgrading"), so this is intentional and documented — flagging it so you make sure release notes/upgrade docs downstream actually surface it prominently, since it's easy to miss in a changelog file.

File: pkg/operatorconfig/operatorconfig.go [hunk 2]
```diff
@@ -35,5 +36,8 @@ func withDefaults(cfg operatorv1.OperatorConfig) operatorv1.OperatorConfig {
 	if cfg.Spec.MaxConcurrentReconciles == 0 {
 		cfg.Spec.MaxConcurrentReconciles = 1
 	}
+	if len(cfg.Spec.WatchedResources) == 0 {
```
Comment (josvazg) [inline pkg/operatorconfig/operatorconfig.go:39]: >3. Minor: empty list vs. omitted field are indistinguishable. if len(cfg.Spec.WatchedResources) == 0 { ... = AllWatchedResources } means a user who explicitly sets watchedResources: [] (intending "watch nothing") silently gets "watch everything" instead. Given the field is x-kubernetes-list-type: set, an empty list is a legal, distinct value from omission. Low-severity edge case, but worth a one-line doc comment or explicit test if "watch nothing" is ever a supported configuration.

File: api/operator/v1/operatorconfig_types.go [hunk 3]
```diff
@@ -25,8 +25,22 @@ const (
...
 	WatchedResourceMongoDBMultiCluster WatchedResource = "mongodbmulticluster"
 	WatchedResourceClusterMongoDBRoles WatchedResource = "clustermongodbroles"
+	WatchedResourceVoyageAI            WatchedResource = "voyageais"
 )
 
+// AllWatchedResources lists every CRD the operator can reconcile. It is the default value
+// for OperatorConfig.spec.watchedResources when the field is omitted.
+var AllWatchedResources = []WatchedResource{
+	WatchedResourceMongoDB,
+	WatchedResourceOpsManagers,
```
Comment (josvazg) [inline api/operator/v1/operatorconfig_types.go:35]: non blocking
> this list now has to stay in sync with the enum marker above and the plural constants in main.go by hand — worth a test that iterates the kubebuilder enum values (or a go:generate) so adding a CRD can't silently miss one of the three spots.

**Review by josvazg — APPROVED** [review]
LGTM with minor comments

File: pkg/operatorconfig/operatorconfig.go (hunk 2 above)
Comment (m1kola) [reply pkg/operatorconfig/operatorconfig.go:39]: "watch nothing" makes no sense for our operator. I don't think it is worth adding complexity to distinguish between empty list and omitted value.

File: main.go (hunk 1 above)
Comment (m1kola) [reply main.go:506]: This is a transient state to keep our E2Es working on the E2E branch. The endgame is to get rid of the ConfigMap and migrate to MemberCluster CRs.

With MemberCluster CRDs our target logic is exactly that - if there are no MemberCluster CRs - we are running in a single-cluster.

**Comment by m1kola** [comment]
> 2. Real upgrade gotcha (documented, but sharp edge): existing OperatorConfig CRs with a restricted watchedResources subset don't get migrated. withDefaults only fills in AllWatchedResources when the field is empty, not merged into an existing non-empty list. So anyone who previously restricted resources via Helm (which never included mongodbmulticluster/clustermongodbroles unless separately opted in) and already has an OperatorConfig CR with an explicit subset will silently lose clustermongodbroles/mongodbmulticluster watching post-upgrade unless they manually update the CR first. The changelog does call this out explicitly ("must create or update their OperatorConfig resource before upgrading"), so this is intentional and documented — flagging it so you make sure release notes/upgrade docs downstream actually surface it prominently, since it's easy to miss in a changelog file.

@josvazg true - this is slightly annoying. But we will:
1. Implement a migration script and it should handle this.
2. We will document in the migration guide for users who won't be using the migration script.

File: docker/mongodb-kubernetes-tests/kubetester/kubetester.py [hunk 4]
```diff
@@ -201,6 +201,7 @@ def apply_operator_config_from_test_env(
     api_client=None,
     name: str = "mongodb-kubernetes-operator",
     wait_for_ready: Optional[Callable[[], None]] = None,
+    extra_spec: Optional[dict] = None,
```
Comment (mircea-cosbuc) [inline docker/mongodb-kubernetes-tests/kubetester/kubetester.py:204]: nit: any reason we're not doing a `**kwargs` pattern, so we can use a call similar to something like this:
```
apply_operator_config_from_test_env(
            self.namespace, api_client=self.api_client, name=self.name, wait_for_ready=wait_for_ready
            self.namespace,
            api_client=self.api_client,
            name=self.name,
            wait_for_ready=wait_for_ready,
            watchedResources={},
            anotherSpecValue=false,
        )
```

File: docker/mongodb-kubernetes-tests/tests/operator/operator_partial_crd.py [hunk 5]
```diff
@@ -24,12 +26,14 @@ def operator_only_ops_manager_and_mongodb(
     namespace: str,
     operator_installation_config: Dict[str, str],
 ) -> Operator:
-    helm_args = operator_installation_config.copy()
-    helm_args["operator.watchedResources"] = "{opsmanagers,mongodb}"
+    # Restrict which CRDs the operator reconciles via OperatorConfig. The CR must exist before the
+    # operator starts: with no CR it defaults to watching all CRDs and would crash on the ones that
+    # aren't installed in this test.
+    create_operator_config(namespace, {"watchedResources": ["opsmanagers", "mongodb"]})
```
Comment (mircea-cosbuc) [inline docker/mongodb-kubernetes-tests/tests/operator/operator_partial_crd.py:32]: It's great to already see this shaping up in the config being equivalent to the helm args.

**Review by mircea-cosbuc — APPROVED** [review]
LGTM

File: docker/mongodb-kubernetes-tests/kubetester/kubetester.py (hunk 4 above)
Comment (m1kola) [reply docker/mongodb-kubernetes-tests/kubetester/kubetester.py:204]: I went with an explicit `extra_spec` dict rather than `**kwargs` because the keys are `OperatorConfig` spec fields (camelCase like `watchedResources`), so a dict mirrors the CR spec 1:1 and keeps those out of the function's snake_case signature. `**kwargs` would also sit next to real params like `multi_cluster` and `wait_for_ready`.
